# Trabajo Final procesamiento de lenguaje natural (NLP) <a class="tocSkip">
## Universidad Pontificia Bolivariana <a class="tocSkip">

**Estudiante:** xxxxxxxxxx

**ID:** xxxxxxxxxx



**Subir Este notebook de Trabajo Final en su repositorio de GitHub. y enviar el link en **Microsoft TEAMS**

Docente: [Jose R. Zapata](https://joserzapata.github.io)
- https://joserzapata.github.io
- https://www.linkedin.com/in/jose-ricardo-zapata-gonzalez/       


## Objetivo del Trabajo
Realizar un proceso de Procesamiento de lenguaje natural (NLP) para dejar los datos preparados para ser usados con algoritmos de Machine Learning para Clasificación como objetivo final del trabajo.

los datos preparados para ser usados con algoritmos de Machine Learning para Regresión o Clasificación como objetivo final del trabajo.

El trabajo se realizara en este jupyter notebook y subirlo a su repositorio de github creado en clase. (**Recuerde poner su nombre e información**)

## Las actividades a realizar
    

1) Limpiar los datos de texto.

   - https://joserzapata.github.io/courses/nlp/procesamiento-basico/

   - https://joserzapata.github.io/courses/nlp/preprocesamiento-texto/

2) Realizar la representación de texto:

   - https://joserzapata.github.io/courses/nlp/representaciones/

   - Tokenización
   - Lematización o stemming
   - Representación de los datos de texto (Bag of Words o TF-IDF)

3) Utilice un modelo de Machine Learning para clasificación para entrenar y evaluar el modelo.
        
    - Calcule estas las métricas de evaluación (accuracy, precision, recall, f1-score) - https://joserzapata.github.io/courses/python-ciencia-datos/clasificacion/#evaluacion-modelo-simple


    
   
    
*NOTA: No dude en contactarme para cualquier pregunta o inquietud :) por el chat de Teams o al correo
joser.zapata@upb.edu.co*

## EVALUACIÓN


|Porcentaje en la evaluación | Descripción| Nada | Incompleto | Completo 
| :---: |:---: |:---: |:---: |:---: 
| 5 % |**limpieza de los datos** |
| 5 % | **Representacion de los datos** |
| 5 % | **Machine Learning** <br> Entrenar y evaluar el modelo propuesto |


In [1]:
import pandas as pd

from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    classification_report,
)
from sklearn.model_selection import train_test_split

Carga del dataset

El archivo  se carga en aproximadamente 10 segundos

In [2]:
# Usar su ID UPB Ejemplo: "0028984798"
id_upb = "000571109"

data_reviews = pd.read_parquet(
    "https://www.dropbox.com/scl/fi/gvk9yj8cn96oocr9z058x/filmaffinity_reviews_dataset.parquet?rlkey=xgvr00zvkxbkwqqavqutpsshg&st=xjb7xze9&dl=1"
)
data_reviews = data_reviews.sample(n=50_000, random_state=int(id_upb))
data_reviews.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50000 entries, 60826 to 17965
Data columns (total 2 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   author_review_desc  50000 non-null  string
 1   author_rating       50000 non-null  int64 
dtypes: int64(1), string(1)
memory usage: 1.1 MB


Ejemplo de algunas filas del dataset

In [3]:
data_reviews.sample(5)

author_review_desc  author_rating
84194   
Nos encontramos ante una bellísima película a...              8
36121   
En cuanto a películas de hombres lobo estamos...              8
42819   
Resumen de la película: tomemos una historia ...              4
123481  
Porque dudo que a muchos de los que la califi...              9
97203   
Es un película simple con personajes arquetíp...              4

Evaluar los valores nulos

In [4]:
data_reviews.isnull().sum()

author_review_desc    0
author_rating         0
dtype: int64

En este caso no hay nulos.

Revisar si hay valores duplicados:

In [5]:
data_reviews.duplicated().sum()

np.int64(3)

No hay  valores duplicados.

## Clasificación Tradicional para Análisis de Sentimientos y Categorías 👍 👎



Definición: Se considera reseña positiva cuando la puntuación ("author_rating") es mayor que 6; negativa en caso contrario.

Crear la variable binaria de sentimiento: 1 (positivo) si author_rating > 6, 0 (negativo) de lo contrario

In [6]:
UMBRAL = 6
data_reviews["sentiment_bin"] = (data_reviews["author_rating"] > UMBRAL).astype(int)

In [7]:
# Dividir los datos en conjuntos de entrenamiento y prueba

X_data = data_reviews["author_review_desc"]
y_data = data_reviews["sentiment_bin"]


Dividir el dataset en entrenamiento y prueba

In [8]:
X_train, X_test, y_train, y_test = train_test_split(
    X_data,
    y_data,
    test_size=0.2,
    stratify=y_data,  # Mantener la proporción de clases en ambos conjuntos
    random_state=42,
)

## Análisis exploratorio de los datos

para determinar que tipo de limpieza se debe realizar a los datos de texto de `X_train`

## 1) Limpieza de los datos de texto

Tomar los datos de `X_train` y aplicar las funciones de limpieza que considere necesarias

Ayudas:

- Convertir a minúsculas
- Eliminar caracteres especiales y números
- Corregir palabras mal escritas
- etc


In [9]:
import re
from tqdm.auto import tqdm

# Configura tqdm para que funcione con los métodos .apply() de pandas
# Esto nos dará una barra de progreso.
tqdm.pandas(desc="Progreso de limpieza")


def limpiar_texto(texto):
    """
    Función para limpiar el texto de las reseñas:
    1. Convierte a minúsculas.
    2. Elimina puntuación, números y caracteres especiales.
    Se preservan caracteres del español (á, é, í, ó, ú, ü, ñ).
    3. Elimina espacios en blanco extra.
    """
    # Asegurarnos de que el input es un string
    if not isinstance(texto, str):
        return ""

    # 1. Convertir a minúsculas
    texto = texto.lower()

    # 2. Eliminar caracteres especiales y números
    # Usamos regex para mantener solo letras (incluyendo acentos y ñ) y espacios
    texto = re.sub(r"[^a-záéíóúüñ\s]", "", texto)

    # 3. Eliminar espacios en blanco extra
    # Reemplaza secuencias de uno o más espacios con un solo espacio
    texto = re.sub(r"\s+", " ", texto).strip()

    return texto


# --- Aplicar la limpieza ---

# (Asumiendo que X_train y X_test ya existen en memoria, como en el notebook)

print(f"Limpiando {len(X_train)} reseñas de X_train...")
# Usamos .progress_apply() gracias a tqdm para ver el avance
X_train_limpio = X_train.progress_apply(limpiar_texto)

print(f"Limpiando {len(X_test)} reseñas de X_test...")
# Es fundamental aplicar EXACTAMENTE la misma limpieza a los datos de prueba
X_test_limpio = X_test.progress_apply(limpiar_texto)

print("\n--- Limpieza completada ---")

# --- Verificar el resultado ---
# Mostramos un ejemplo antes y después
# (Usamos .iloc[0] para obtener la primera fila de la Serie de entrenamiento)
print("\nEJEMPLO DE LIMPIEZA:")
print("------------------------")
print("RESEÑA ORIGINAL (X_train):")
print(X_train.iloc[0])
print("------------------------")
print("RESEÑA LIMPIA (X_train_limpio):")
print(X_train_limpio.iloc[0])


Limpiando 40000 reseñas de X_train...


Progreso de limpieza:   0%|          | 0/40000 [00:00<?, ?it/s]

Limpiando 10000 reseñas de X_test...


Progreso de limpieza:   0%|          | 0/10000 [00:00<?, ?it/s]


--- Limpieza completada ---

EJEMPLO DE LIMPIEZA:
------------------------
RESEÑA ORIGINAL (X_train):

El argumento es claro, hay algo más allá de la muerte. Para ello el maestro nos introduce la historia de tres personajes, el primero, Damon, alguien con el don o la maldición de poder poner en contacto a vivos y muertos, hecho que le atormenta. El segundo, una preciosa Célice de France, famosa reportera de televisión, que tras llegar al abismo de la muerte, parece que empieza a reflexionar acerca de la experiencia que ha tenido. Finalmente, McLaren, un tímido niño inseparable de su gemelo que sufre una experiencia terrible que le marcará siempre.


Hasta ahí todo vale, y la película es entretenida, y en mi opinión nunca aburre. Bien es cierto que tiene momentos más brillantes que otros, sin embargo, y como indica Kirk Honeycutt (The Hollywood Reporter), el final decepciona. Pese a que el hilo narrativo es correcto falta algo en la trama que efectivamente nos haga vibrar de verdad.




## 2) Representacion del texto

Luego de tener los datos limpios, realizar la representación de los datos de texto para poder usarse en modelos de machine learning.

In [10]:
import nltk
from nltk.corpus import stopwords
from nltk.stem import SnowballStemmer
from sklearn.feature_extraction.text import TfidfVectorizer

# --- 1. Preparación de Stopwords y Stemmer ---

# Descargar los recursos de NLTK (stopwords) si no están presentes
try:
    nltk.data.find("corpora/stopwords")
    nltk.data.find("tokenizers/punkt")  # El stemmer lo puede necesitar
except LookupError:
    print("Descargando recursos de NLTK (stopwords, punkt)...")
    nltk.download("stopwords")
    nltk.download("punkt")
    print("Recursos descargados.")

# Cargar stopwords en español
spanish_stopwords = list(stopwords.words("spanish"))
print(f"Se cargaron {len(spanish_stopwords)} stopwords en español.")
print(f"Ejemplos: {spanish_stopwords[::50]}")  # Muestra algunas

# Instanciar el Stemmer en español
stemmer = SnowballStemmer("spanish")

# --- 2. Definición del Tokenizador con Stemming ---

# TfidfVectorizer no tiene un parámetro 'stemmer',
# pero podemos pasarle una función 'tokenizer' personalizada.

# Primero, obtenemos el tokenizador por defecto de scikit-learn
# (Este sabe cómo manejar puntuación y espacios que el Paso 1 pudo dejar)
default_tokenizer = TfidfVectorizer().build_tokenizer()


def stemmed_tokenizer(doc):
    """
    Función que primero tokeniza (separa en palabras) y luego
    aplica stemming a cada token.
    """
    # Usamos el tokenizador por defecto
    tokens = default_tokenizer(doc)
    # Aplicamos stemming a cada token
    stemmed_tokens = [stemmer.stem(token) for token in tokens]
    return stemmed_tokens


print("\nTokenizador con stemming configurado.")
print("Ejemplo de stemming:")
print(f"  'películas', 'película', 'peliculón' -> '{stemmer.stem('películas')}'")


# --- 3. Configuración del Vectorizador TF-IDF ---

vectorizer = TfidfVectorizer(
    tokenizer=stemmed_tokenizer,  # Usar nuestra función personalizada
    stop_words=spanish_stopwords,  # Eliminar palabras comunes
    ngram_range=(1, 2),  # Considerar palabras solas (1-gram) y pares (2-gram)
    max_features=5000,  # Limitar vocabulario a las 5000 features más relevantes
    min_df=3,  # Ignorar palabras/n-grams que aparezcan en < 3 reseñas
)

print("\nVectorizador TF-IDF configurado.")

# --- 4. Aplicar la representación (Fit y Transform) ---

# (Asumiendo que X_train_limpio y X_test_limpio existen del Paso 1)

print("Iniciando vectorización...")

# 1. Aprender el vocabulario (fit) y transformar los datos de ENTRENAMIENTO
print("Ajustando (fit) el vectorizador y transformando X_train...")
# X_train_limpio y y_train (del script original)
X_train_tfidf = vectorizer.fit_transform(X_train_limpio)

# 2. SOLO transformar los datos de PRUEBA (usando el vocabulario aprendido)
print("Transformando X_test...")
# X_test_limpio (del script original)
X_test_tfidf = vectorizer.transform(X_test_limpio)

print("\n--- Representación completada ---")

# --- 5. Verificar el resultado ---
print(f"\nForma (shape) de la matriz X_train_tfidf: {X_train_tfidf.shape}")
print(f"Forma (shape) de la matriz X_test_tfidf: {X_test_tfidf.shape}")

print("\nExplicación de la forma:")
print(f" - {X_train_tfidf.shape[0]} filas: (número de reseñas en X_train)")
print(f" - {X_train_tfidf.shape[1]} columnas: (el vocabulario de max_features)")

# Opcional: Mostrar algunas features (palabras/n-grams) aprendidas
try:
    feature_names = vectorizer.get_feature_names_out()
    print("\nAlgunas features (palabras/n-grams) aprendidas por el vectorizador:")
    # Imprime una de cada 500 features para dar una idea
    print(list(feature_names[::500]))
except Exception as e:
    print(f"No se pudieron mostrar features: {e}")


Descargando recursos de NLTK (stopwords, punkt)...


[nltk_data] Downloading package stopwords to /home/vscode/nltk_data...
[nltk_data]   Unzipping corpora/stopwords.zip.
[nltk_data] Downloading package punkt to /home/vscode/nltk_data...
[nltk_data]   Unzipping tokenizers/punkt.zip.


Recursos descargados.
Se cargaron 313 stopwords en español.
Ejemplos: ['de', 'contra', 'tuya', 'estuviera', 'hubieras', 'fuerais', 'tuvierais']

Tokenizador con stemming configurado.
Ejemplo de stemming:
  'películas', 'película', 'peliculón' -> 'pelicul'

Vectorizador TF-IDF configurado.
Iniciando vectorización...
Ajustando (fit) el vectorizador y transformando X_train...


/workspaces/NLP_PabloZapataOchoa/.venv/lib/python3.12/site-packages/sklearn/feature_extraction/text.py:517: UserWarning: The parameter 'token_pattern' will not be used since 'tokenizer' is not None'
  warnings.warn(
/workspaces/NLP_PabloZapataOchoa/.venv/lib/python3.12/site-packages/sklearn/feature_extraction/text.py:402: UserWarning: Your stop_words may be inconsistent with your preprocessing. Tokenizing the stop words generated tokens ['algun', 'com', 'contr', 'cuand', 'desd', 'dond', 'durant', 'eram', 'estab', 'estais', 'estam', 'estan', 'estand', 'estaran', 'estaras', 'esteis', 'estem', 'esten', 'estes', 'estuv', 'fuer', 'fues', 'fuim', 'fuist', 'hab', 'habr', 'habran', 'habras', 'hast', 'hem', 'hub', 'mas', 'mia', 'mias', 'mio', 'mios', 'much', 'nad', 'nosotr', 'nuestr', 'par', 'per', 'poc', 'porqu', 'qui', 'seais', 'seam', 'sent', 'ser', 'seran', 'seras', 'si', 'sient', 'sint', 'sobr', 'som', 'suy', 'tambien', 'tant', 'ten', 'tendr', 'tendran', 'tendras', 'teng', 'tien', 'tod', '

Transformando X_test...

--- Representación completada ---

Forma (shape) de la matriz X_train_tfidf: (40000, 5000)
Forma (shape) de la matriz X_test_tfidf: (10000, 5000)

Explicación de la forma:
 - 40000 filas: (número de reseñas en X_train)
 - 5000 columnas: (el vocabulario de max_features)

Algunas features (palabras/n-grams) aprendidas por el vectorizador:
['abaj', 'blanchett', 'corean', 'ello', 'grup', 'lagun', 'nazis', 'per ser', 'reserv', 'tant pelicul']


## 3) Entrenar un modelo de machine learning de clasificación

Utilizar un modelo de clasificación para entrenar y evaluar el modelo con los datos preparados.


In [11]:
import joblib
from sklearn.linear_model import LogisticRegression
from sklearn.metrics import (
    accuracy_score,
    precision_score,
    recall_score,
    f1_score,
    roc_auc_score,
    classification_report,
)

# (Asumimos que las variables X_train_tfidf, y_train, X_test_tfidf, y_test
#  existen en memoria desde los pasos anteriores)

print("--- Iniciando Paso 3: Entrenamiento del Modelo ---")

# --- 1. Instanciar el Modelo ---
# Elegimos LogisticRegression. Es un modelo lineal robusto y rápido,
# ideal para datos TF-IDF que son "dispersos" (muchos ceros).
# - random_state=42: Asegura que obtengamos los mismos resultados cada vez.
# - max_iter=1000: Aumentamos las iteraciones para asegurar que el modelo
#   "converja" (encuentre la mejor solución) con nuestros 5000 features.
model = LogisticRegression(random_state=42, max_iter=1000)

print(f"Modelo seleccionado: {type(model).__name__}")

# --- 2. Entrenar el Modelo ---
print("\nEntrenando el modelo con los datos de X_train_tfidf...")
# Usamos .fit() para que el modelo aprenda la relación entre
# la matriz TF-IDF (las palabras) y el sentimiento (0 o 1).
model.fit(X_train_tfidf, y_train)

print("¡Modelo entrenado exitosamente!")

# --- 3. Evaluar el modelo con los datos de prueba ---

print("\n--- Iniciando Evaluación del Modelo ---")
print("Generando predicciones en el conjunto de prueba (X_test_tfidf)...")

# Usamos .predict() en los datos de prueba (X_test_tfidf)
# El modelo NUNCA vio estos datos durante el entrenamiento.
y_pred = model.predict(X_test_tfidf)

# Opcional: Calcular probabilidades para ROC-AUC
# (Mide qué tan bien distingue el modelo entre clases)
y_pred_proba = model.predict_proba(X_test_tfidf)[:, 1]

print("Predicciones generadas.")

# --- 4. Calcular y Reportar Métricas ---
# Comparamos las predicciones (y_pred) con las etiquetas reales (y_test)

print("\n--- Métricas de Evaluación ---")

# Accuracy: ¿Qué porcentaje de reseñas clasificó correctamente?
# (Correctas / Total)
acc = accuracy_score(y_test, y_pred)
print(f"Accuracy (Exactitud): {acc:.4f}")

# Precision: De todas las que el modelo dijo que eran "Positivas",
# ¿cuántas realmente lo eran?
# (Importante si queremos evitar "Falsos Positivos")
prec = precision_score(y_test, y_pred)
print(f"Precision: {prec:.4f}")

# Recall (Sensibilidad): De todas las reseñas que *eran* "Positivas",
# ¿cuántas logró encontrar el modelo?
# (Importante si queremos encontrar todos los "Positivos Reales")
rec = recall_score(y_test, y_pred)
print(f"Recall (Sensibilidad): {rec:.4f}")

# F1-Score: La media armónica entre Precision y Recall.
# Un buen balance general entre ambas.
f1 = f1_score(y_test, y_pred)
print(f"F1 Score: {f1:.4f}")

# ROC-AUC: Mide la habilidad del modelo para discriminar
# entre la clase positiva y negativa. (1.0 es perfecto, 0.5 es azar)
roc_auc = roc_auc_score(y_test, y_pred_proba)
print(f"ROC AUC Score: {roc_auc:.4f}")

# Reporte de Clasificación: Un resumen completo
print("\n--- Reporte de Clasificación Detallado ---")
# Muestra Precision, Recall y F1 para ambas clases (0 y 1)
print(
    classification_report(y_test, y_pred, target_names=["Negativo (0)", "Positivo (1)"])
)


# --- 5. Guardar el Modelo Entrenado y el Vectorizador ---
# Es crucial guardar también el 'vectorizer' del Paso 2,
# porque para predecir en nuevos datos, necesitamos aplicar
# EXACTAMENTE la misma transformación TF-IDF.

print("\n--- Guardando artefactos del modelo ---")

# Guardamos el modelo entrenado
joblib.dump(model, "modelo_sentimientos.joblib")
print("Modelo guardado en 'modelo_sentimientos.joblib'")

# (Asumiendo que 'vectorizer' es el nombre de tu TfidfVectorizer del Paso 2)
try:
    joblib.dump(vectorizer, "vectorizador_tfidf.joblib")
    print("Vectorizador guardado en 'vectorizador_tfidf.joblib'")
except NameError:
    print("ADVERTENCIA: No se pudo guardar el 'vectorizer'.")
    print("Asegúrate de que la variable 'vectorizer' (del Paso 2) esté en memoria.")
except Exception as e:
    print(f"Error al guardar el vectorizador: {e}")

print("\n--- Proceso completado ---")


--- Iniciando Paso 3: Entrenamiento del Modelo ---
Modelo seleccionado: LogisticRegression

Entrenando el modelo con los datos de X_train_tfidf...
¡Modelo entrenado exitosamente!

--- Iniciando Evaluación del Modelo ---
Generando predicciones en el conjunto de prueba (X_test_tfidf)...
Predicciones generadas.

--- Métricas de Evaluación ---
Accuracy (Exactitud): 0.8020
Precision: 0.8037
Recall (Sensibilidad): 0.8345
F1 Score: 0.8188
ROC AUC Score: 0.8837

--- Reporte de Clasificación Detallado ---
              precision    recall  f1-score   support

Negativo (0)       0.80      0.76      0.78      4639
Positivo (1)       0.80      0.83      0.82      5361

    accuracy                           0.80     10000
   macro avg       0.80      0.80      0.80     10000
weighted avg       0.80      0.80      0.80     10000


--- Guardando artefactos del modelo ---
Modelo guardado en 'modelo_sentimientos.joblib'
Vectorizador guardado en 'vectorizador_tfidf.joblib'

--- Proceso completado ---


Métricas de Evaluación Clave:

Accuracy (Exactitud): 0.8020 (80.2%)

El modelo clasificó correctamente 8 de cada 10 reseñas en el conjunto de prueba. Este es un rendimiento sólido y robusto.

ROC AUC Score: 0.8837 (88.4%)

Esta métrica, que mide la capacidad del modelo para discriminar entre una reseña positiva y una negativa, es muy alta. Un valor de 0.88 (cercano a 1.0) indica que el modelo es muy confiable para distinguir las clases.

Reporte de Clasificación (F1-Score):

Positivo (1): El modelo obtuvo un F1-Score de 0.82.

Negativo (0): El modelo obtuvo un F1-Score de 0.78.

El rendimiento es balanceado (Macro Avg F1-Score: 0.80), lo que demuestra que el modelo no tiene un sesgo significativo hacia una clase y es competente identificando tanto reseñas positivas como negativas.

Conclusión Final:

El pipeline de NLP clásico (Limpieza -> Stemming -> TF-IDF) combinado con un modelo lineal (Regresión Logística) ha demostrado ser muy efectivo, alcanzando más de un 80% de exactitud. Los artefactos del modelo (modelo_sentimientos.joblib) y el vectorizador (vectorizador_tfidf.joblib) se han guardado exitosamente y están listos para ser desplega

### Evaluar el modelo con los datos de prueba

Usar el modelo para predecir en `X_test` y evaluar con `y_test`

**Nota:** Recuerde que `X_test` debe pasar por los mismos procesos de limpieza y representación que `X_train`

In [ ]:
print("Accuracy:", accuracy_score(y_test, y_pred))
print("Precision:", precision_score(y_test, y_pred))
print("Recall:", recall_score(y_test, y_pred))
print("F1 Score:", f1_score(y_test, y_pred))
print("ROC AUC Score:", roc_auc_score(y_test, y_pred))
print("Reporte de clasificación:\n", classification_report(y_test, y_pred))

**Guardar el modelo**

In [ ]:
import joblib

## Librerías Usadas

In [ ]:
from watermark import watermark

print(watermark(python=True, iversions=True, globals_=globals()))

---

## Referencias
- [CountVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.CountVectorizer.html)
- [TfidfVectorizer](https://scikit-learn.org/stable/modules/generated/sklearn.feature_extraction.text.TfidfVectorizer.html)
- [Classification of text documents using sparse features](https://scikit-learn.org/stable/auto_examples/text/plot_document_classification_20newsgroups.html)
- Ejemplo de Entrenamiento y selección de Modelo de machine learning entre varios modelos <https://joserzapata.github.io/post/ciencia-datos-proyecto-python/6-model_selection/>
- https://joserzapata.github.io/courses/python-ciencia-datos/python/
- https://joserzapata.github.io/courses/python-ciencia-datos/pandas/
- https://joserzapata.github.io/courses/python-ciencia-datos/machine-learning/

Docente: [Jose R. Zapata](https://joserzapata.github.io)

- [https://joserzapata.github.io/](https://joserzapata.github.io/)
- [https://www.linkedin.com/in/jose-ricardo-zapata-gonzalez/](https://www.linkedin.com/in/jose-ricardo-zapata-gonzalez/)